### PR TITLE
[IMP] topbar: allow to insert checkbox in a cell

### PR DESCRIPTION
**Pull request by anhe-odoo**
## Task Description

Now that we have the datavalidation, this PR aims to add a simple item in the Insert menu to let the user directly insert a checkbox in a cell without having to use the Data-validation side-pannel. To mimic Google sheet behavior, we are now also able to remove a checkbox from a cell in the same way we can delete the content of a cell (e.g. with 'delete' or 'backspace' key).

## Related Task
- Task: [3175856](https://www.odoo.com/web#id=3175856&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/actions/menu_items_actions.ts [hunk 1]
```diff
@@ -436,10 +436,27 @@ export const OPEN_CF_SIDEPANEL_ACTION = (env: SpreadsheetChildEnv) => {
 };
 
 export const INSERT_LINK = (env: SpreadsheetChildEnv) => {
-  let { col, row } = env.model.getters.getActivePosition();
+  const { col, row } = env.model.getters.getActivePosition();
   env.model.dispatch("OPEN_CELL_POPOVER", { col, row, popoverType: "LinkEditor" });
 };
 
+export const INSERT_CHECKBOX = (env: SpreadsheetChildEnv) => {
```
Comment (LucasLefevre) [inline src/actions/menu_items_actions.ts:443]: you can inline the implementation if it's only used once.

File: src/actions/menu_items_actions.ts [hunk 2]
```diff
@@ -436,10 +436,27 @@ export const OPEN_CF_SIDEPANEL_ACTION = (env: SpreadsheetChildEnv) => {
...
+    sheetId,
+    rule: {
+      id: env.model.uuidGenerator.uuidv4(),
+      criterion: {
+        type: "isBoolean",
+        values: [],
+      },
+    },
+  });
+  env.model.dispatch("UPDATE_CELL", { sheetId, col, row, content: "FALSE" });
```
Comment (LucasLefevre) [inline src/actions/menu_items_actions.ts:457]: two dispatches means two history steps (you would need to hit Ctrl+Z twice to undo the change). We might need to add a new "helper" command to do that. e.g. INSERT_CHECKBOX

File: src/plugins/core/data_validation.ts [hunk 3]
```diff
@@ -122,6 +122,24 @@ export class DataValidationPlugin
         this.addDataValidationRule(cmd.sheetId, { ...cmd.rule, ranges });
         break;
       }
+      case "DELETE_CONTENT": {
```
Comment (LucasLefevre) [inline src/plugins/core/data_validation.ts:125]: what about an UPDATE_CELL with `content: ""` ?

**Review by LucasLefevre — COMMENTED** [review]
yop, quick look :)

File: src/plugins/core/data_validation.ts (hunk 3 above)
Comment (anhe-odoo) [reply src/plugins/core/data_validation.ts:125]: It's a strange case in fact. For example, selecting the cell and emptying the composer should do nothing to the checkbox, but copying an empty cell over the "checkboxed" cell should clear it, so we should have a way to differentiate these things

Comment (LucasLefevre) [reply src/plugins/core/data_validation.ts:125]: ok, fair if it's done on purpose :+1: 
Is it tested ?

File: src/actions/menu_items_actions.ts (hunk 2 above)
Comment (anhe-odoo) [reply src/actions/menu_items_actions.ts:457]: After a discussion with Adrien, we decided to change the way Checkbox is added in the datavalidation rule and always add "FALSE" as the default content

File: src/actions/insert_actions.ts [hunk 4]
```diff
@@ -248,6 +248,26 @@ export const insertLink: ActionSpec = {
   icon: "o-spreadsheet-Icon.INSERT_LINK",
 };
 
+export const insertCheckbox: ActionSpec = {
+  name: _t("Checkbox"),
+  execute: (env) => {
+    const { sheetId, col, row } = env.model.getters.getActivePosition();
+    const ranges = [env.model.getters.getRangeDataFromXc(sheetId, toXC(col, row))];
```
Comment (hokolomopo) [inline src/actions/insert_actions.ts:255]: should be the whole selected range, not only the anchor cell no ? At least it work like that in GSheet

File: src/helpers/clipboard/clipboard_cells_state.ts [hunk 5]
```diff
@@ -627,6 +627,10 @@ export class ClipboardCellsState extends ClipboardCellsAbstractState {
   private pasteDataValidation(origin: CellPosition, target: CellPosition) {
     const rule = this.getters.getValidationRuleForCell(origin);
     if (!rule) {
+      const targetRul = this.getters.getValidationRuleForCell(target);
```
Comment (hokolomopo) [inline src/helpers/clipboard/clipboard_cells_state.ts:630]:
```suggestion
      const targetRule = this.getters.getValidationRuleForCell(target);
```

File: src/plugins/core/data_validation.ts [hunk 6]
```diff
@@ -122,6 +124,24 @@ export class DataValidationPlugin
...
+        const zones = cmd.target;
+        const sheetId = cmd.sheetId;
+        for (const zone of zones) {
+          for (let row = zone.top; row <= zone.bottom; row++) {
+            for (let col = zone.left; col <= zone.right; col++) {
+              const dataValidation = this.getValidationRuleForCell({ sheetId, col, row });
+              if (!dataValidation || dataValidation.criterion.type !== "isBoolean") {
```
Comment (hokolomopo) [inline src/plugins/core/data_validation.ts:134]: functionally is kinda strange that pressing "Delete" removes the checkboxes but not any other DV rule...

File: src/plugins/core/data_validation.ts [hunk 7]
```diff
@@ -192,16 +212,19 @@ export class DataValidationPlugin
...
       const cell = this.getters.getCell(position);
-      const { sheetId, col, row } = position;
       const style: Style = {
         ...cell?.style,
         align: cell?.style?.align ?? "center",
         verticalAlign: cell?.style?.verticalAlign ?? "middle",
       };
-      this.dispatch("UPDATE_CELL", { sheetId, col, row, style });
+      let updatedCell: Omit<UpdateCellCommand, "type"> = { ...position, style };
+      if (cell?.content !== "true" && cell?.content !== "TRUE") {
```
Comment (hokolomopo) [inline src/plugins/core/data_validation.ts:224]:
```suggestion
      if (cell?.content.toLowerCase() !== "true") {
```
more robust if it's ever possible to have cell values be somthing like `TrUe`

File: tests/data_validation/data_validation_checkbox_component.test.ts [hunk 8]
```diff
@@ -28,6 +33,24 @@ describe("Checkbox in model", () => {
     });
   });
 
+  test("Checkbox are removed when clearing the content of the cell", () => {
```
Comment (hokolomopo) [inline tests/data_validation/data_validation_checkbox_component.test.ts:36]: Also test that other rule types aren't removed if it's intended to work like that

File: src/helpers/clipboard/clipboard_cells_state.ts [hunk 9]
```diff
@@ -627,6 +627,10 @@ export class ClipboardCellsState extends ClipboardCellsAbstractState {
...
+      if (targetRul) {
+        this.adaptDataValidationRule(target.sheetId, targetRul, [], [toXC(target.col, target.row)]);
```
Comment (hokolomopo) [inline src/helpers/clipboard/clipboard_cells_state.ts:632]: Nitpick: i'd add a comment that we're removing the DV rule on the target, kinda hard to see what we're doing by reading the code normally. adaptDataValidationRule are kinda obscure ^^'

**Review by hokolomopo — COMMENTED** [review]
Inserting checkboxes on a range doesn't work, otherwise some details and it looks fine :)

File: src/plugins/core/data_validation.ts (hunk 6 above)
Comment (anhe-odoo) [reply src/plugins/core/data_validation.ts:134]: Gsheet has the same behavior except for the combobox, which should probably be insertable in the same way in another task, so I though we should only keep the checkbox for the moment, and see for the combobox when it will be insertable :)

File: src/helpers/clipboard/clipboard_cells_state.ts [hunk 10]
```diff
@@ -627,6 +627,16 @@ export class ClipboardCellsState extends ClipboardCellsAbstractState {
   private pasteDataValidation(origin: CellPosition, target: CellPosition) {
     const rule = this.getters.getValidationRuleForCell(origin);
     if (!rule) {
+      const targetRule = this.getters.getValidationRuleForCell(target);
```
Comment (rrahir) [inline src/helpers/clipboard/clipboard_cells_state.ts:630]: Is this related to the new feature? I feel like it's more of a bug fix or maybe the situation could not occur beforehand?

File: src/plugins/core/data_validation.ts (hunk 6 above)
Comment (rrahir) [reply src/plugins/core/data_validation.ts:134]: imo, the data validation update upon "delete content" is kind of unrelated to the insertion. If anything it is something that was overlooked in the first implementation and should be added in parallel, but not type by type

File: src/plugins/core/data_validation.ts [hunk 11]
```diff
@@ -192,16 +212,19 @@ export class DataValidationPlugin
...
       const cell = this.getters.getCell(position);
-      const { sheetId, col, row } = position;
       const style: Style = {
         ...cell?.style,
         align: cell?.style?.align ?? "center",
         verticalAlign: cell?.style?.verticalAlign ?? "middle",
       };
-      this.dispatch("UPDATE_CELL", { sheetId, col, row, style });
+      let updatedCell: Omit<UpdateCellCommand, "type"> = { ...position, style };
+      if (cell?.content.toLowerCase() !== "true") {
```
Comment (rrahir) [inline src/plugins/core/data_validation.ts:224]: the data validation also works for the evaluated functions that return a boolean value.
 `A1:-> =1=1` should be ticked by default

File: src/helpers/clipboard/clipboard_cells_state.ts (hunk 10 above)
Comment (anhe-odoo) [reply src/helpers/clipboard/clipboard_cells_state.ts:630]: It's not really a bug fix but a change in the desired behavior we discussed with Francois. But it should probably be added in a separated commit

File: src/components/side_panel/data_validation/data_validation_panel.ts [hunk 12]
```diff
@@ -19,6 +20,21 @@ export class DataValidationPanel extends Component<Props, SpreadsheetChildEnv> {
...
+  setup(): void {
+    if (this.props.ruleID) {
+      this.onPreviewClick(this.props.ruleID);
+    }
+    onWillUpdateProps((nextProps: Props) => {
+      if (nextProps.ruleID !== this.props.ruleID) {
+        if (nextProps.ruleID) {
+          this.onPreviewClick(nextProps.ruleID);
+        } else {
+          this.onExitEditMode();
```
Comment (LucasLefevre) [inline src/components/side_panel/data_validation/data_validation_panel.ts:32]: What about adding another entry in the side panel registry to be able to open the DV "editor" directly ?
`onWillUpdateProps` are always smelly ;)
It would also keep this component simpler

**Review by LucasLefevre — COMMENTED** [review]
quick suggestion ;)

File: src/components/side_panel/data_validation/data_validation_panel.ts (hunk 12 above)
Comment (anhe-odoo) [reply src/components/side_panel/data_validation/data_validation_panel.ts]: Done :)

File: src/registries/side_panel_registry.ts [hunk 13]
```diff
@@ -62,6 +63,11 @@ sidePanelRegistry.add("DataValidation", {
   Body: DataValidationPanel,
 });
 
+sidePanelRegistry.add("DataValidationEditor", {
+  title: _t("Data validation editor"),
```
Comment (LucasLefevre) [inline src/registries/side_panel_registry.ts:67]: I would have the same title. Such that the title is the same no matter how this side panel is open
```suggestion
  title: _t("Data validation"),
```

File: tests/menus/menu_items_registry.test.ts [hunk 14]
```diff
@@ -929,6 +929,59 @@ describe("Menu Item actions", () => {
     restoreDefaultFunctions();
   });
 
+  describe("Insert -> Checkbox", () => {
+    test("In an empty cell", () => {
```
Comment (LucasLefevre) [inline tests/menus/menu_items_registry.test.ts:933]: I'd move all those tests with the DV plugins tests (either core or evaluation, there's no obvious place since it involves both). Most of them are testing edge cases which are handled inside the plugins, and they are not specific to the menu item in the registry.
So instead of `doAction` => use `addDataValidation` like other tests in those files.

You can leave only one simple test case here to test the menu item itself.

EDIT: actually, the file `data_validation_checkbox_plugin.test.ts` is the perfect file ;)

Also, the file `data_validation_checkbox_component.test.ts` does not test the component and seem a vague copy-pasted version of the other file.

File: tests/data_validation/data_validation_checkbox_component.test.ts (hunk 8 above)
Comment (LucasLefevre) [inline tests/data_validation/data_validation_checkbox_component.test.ts:36]: move those to `data_validation_checkbox_plugin.test.ts`

File: tests/data_validation/data_validation_list_component.test.ts [hunk 15]
```diff
@@ -233,6 +238,32 @@ describe("autocomplete in composer", () => {
...
       expect(model.getters.getEditionMode()).toBe("inactive");
     });
+
+    test("Dropdown lists are kept when clearing the content of a non-empty cell", () => {
+      expect(model.getters.getDataValidationListCellsPositions()).toEqual([
+        { sheetId: model.getters.getActiveSheetId(), col: 0, row: 0 },
+      ]);
+      setCellContent(model, "A1", "hello");
+      deleteContent(model, ["A1"]);
+      expect(model.getters.getDataValidationListCellsPositions()).toHaveLength(1);
```
Comment (LucasLefevre) [inline tests/data_validation/data_validation_list_component.test.ts:248]: you can check that "hello" was cleared

File: tests/data_validation/data_validation_list_component.test.ts [hunk 16]
```diff
@@ -233,6 +238,32 @@ describe("autocomplete in composer", () => {
...
+    test("Dropdown lists are removed when clearing the content of an empty cell", () => {
+      expect(model.getters.getDataValidationListCellsPositions()).toEqual([
+        { sheetId: model.getters.getActiveSheetId(), col: 0, row: 0 },
+      ]);
+      setCellContent(model, "A1", "");
+      deleteContent(model, ["A1"]);
+      expect(model.getters.getDataValidationListCellsPositions()).toHaveLength(0);
+    });
+
+    test("Dropdown lists are removed when clearing the content of an empty cell", () => {
```
Comment (LucasLefevre) [inline tests/data_validation/data_validation_list_component.test.ts:260]: this test title is the same as the test above. So I'm not sure what it's supposed to test

**Review by LucasLefevre — COMMENTED** [review]
only minor suggestions :)

File: src/plugins/core/data_validation.ts [hunk 17]
```diff
@@ -122,6 +123,31 @@ export class DataValidationPlugin
...
+          for (let row = zone.top; row <= zone.bottom; row++) {
+            for (let col = zone.left; col <= zone.right; col++) {
+              const dataValidation = this.getValidationRuleForCell({ sheetId, col, row });
+              if (
+                !dataValidation ||
+                (dataValidation.criterion.type !== "isBoolean" &&
+                  !(
+                    dataValidation.criterion.type === "isValueInList" &&
+                    !this.getters.getCell({ sheetId, col, row })?.content
+                  ))
```
Comment (LucasLefevre) [inline src/plugins/core/data_validation.ts:139]: maybe written the other way around, the condition would be easier to read ?

File: tests/data_validation/data_validation_list_component.test.ts [hunk 18]
```diff
@@ -233,6 +238,33 @@ describe("autocomplete in composer", () => {
       expect(getCellContent(model, "A1")).toBe("hello");
       expect(model.getters.getEditionMode()).toBe("inactive");
     });
+
+    test("Dropdown lists are kept when clearing the content of a non-empty cell", () => {
```
Comment (LucasLefevre) [inline tests/data_validation/data_validation_list_component.test.ts:242]: those tests could be move elsewhere than this file (component tests). I guess there are already other plugin tests for lists

Comment (anhe-odoo) [reply tests/data_validation/data_validation_list_component.test.ts:242]: While there were a plugin for the checkbox, I don't think there is any plugin tests for dropdown list, I'll create one :)

**Review by LucasLefevre — APPROVED** [review]
robodoo r+
